framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -407,71 +406,47 @@ public S initialStore(UnderlyingAST underlyingAST, @Nullable List<LocalVariableN
     return info;
   }
 
-  private void addFieldValues(
-      S info, AnnotatedTypeFactory factory, ClassTree classTree, MethodTree methodTree) {
-
-    // Add knowledge about final fields, or values of non-final fields
-    // if we are inside a constructor (information about initializers)
-    TypeMirror classType = TreeUtils.typeOf(classTree);
-    List<Pair<VariableElement, V>> fieldValues = analysis.getFieldValues();
-    for (Pair<VariableElement, V> p : fieldValues) {
-      VariableElement element = p.first;
-      V value = p.second;
-      if (ElementUtils.isFinal(element) || TreeUtils.isConstructor(methodTree)) {
-        JavaExpression receiver;
-        if (ElementUtils.isStatic(element)) {
-          receiver = new ClassName(classType);
-        } else {
-          receiver = new ThisReference(classType);
-        }
-        TypeMirror fieldType = ElementUtils.getType(element);
-        JavaExpression field = new FieldAccess(receiver, fieldType, element);
-        info.insertValue(field, value);
+  /**
+   * Add field values to the initial store before {@code methodTree}.
+   *
+   * <p>The initializer value is inserted into {@code info} if the field is private and final.
+   *
+   * <p>If {@code methodTree} is a constructor, then the declared value is inserted into {@code
+   * info} if the field has an initializer. (Use the declaration type rather than the initializer
+   * because an initialization block might have re-set it.) If {@code methodTree} is not a
+   * constructor, then the declared value is inserted into {@code info} if the receiver of the
+   * method is fully initialized as determined by {@link
+   * #isNotFullyInitializedReceiver(MethodTree)}.
+   *
+   * @param info initial store into which field values are inserted
+   * @param classTree the class that contains {@code methodTree}
+   * @param methodTree the method or constructor tree
+   */
+  private void addFieldValues(S info, ClassTree classTree, MethodTree methodTree) {
+    boolean isConstructor = TreeUtils.isConstructor(methodTree);
+    List<FieldValues<V>> fields = analysis.getFieldValues();
+    TypeElement classEle = TreeUtils.elementFromDeclaration(classTree);
+    for (FieldValues<V> fieldValues : fields) {
```

Comment:
I think it would be clearer to name the first variable `fieldValues` and the last variable `fieldValue`.  That is a motivation for changing the class name `FieldValues` to a singular one.